Added/updated tests for matchAll

peterwmwong opened the pull request:
Tests were updated and assume https://github.com/tc39/proposal-string-matchall/pull/33 will be merged.

In an attempt to make it easier to see the coverage these tests provides, here's a version of the specification annotated with links to the related test from this PR: https://gist.github.com/peterwmwong/2a9d80d3533c0ab165bc8c675ac99e45 (Example below)

<img width="571" alt="screen shot 2018-03-18 at 7 53 35 pm" src="https://user-images.githubusercontent.com/284734/37573460-1fb56f98-2ae6-11e8-8483-93a1f05d66e3.png">

I haven't had the honor of contributing to Test262, so I'm not sure if I've made some bad cross cutting decisions...
- `const`/`let` over `var`
- Using arrow function
- Frontmatter `features`, I'm adding `String.prototype.matchAll` and `Symbol.matchAll`

@ljharb I totally defer to you as you're the expert, let me know whatever I can do to make these tests better!

Thanks!

PS - If you'd like to try out my in-progress V8 implementation, checkout my [branch](https://github.com/peterwmwong/v8/tree/string-matchAll) or download the pre-built MacOS X binary here (zip): https://gist.github.com/peterwmwong/e38b50689d7717870b2cff7b68574b87

peterwmwong commented on test/built-ins/RegExp/prototype/Symbol.matchAll/get-species-constructor-err.js:
```diff
@@ -1,22 +0,0 @@
-// Copyright (C) 2015 the V8 project authors. All rights reserved.
```
This test is now: [species-constructor-get-species-throws.js](https://github.com/peterwmwong/test262/blob/matchall-pww/test/built-ins/RegExp/prototype/Symbol.matchAll/species-constructor-get-species-throws.js)

peterwmwong commented on test/built-ins/RegExp/prototype/Symbol.matchAll/this-val-non-obj.js:
```diff
@@ -1,36 +0,0 @@
-// Copyright (C) 2018 Jordan Harband. All rights reserved.
```
This test is now: [this-not-object-throws.js](https://github.com/peterwmwong/test262/blob/matchall-pww/test/built-ins/RegExp/prototype/Symbol.matchAll/this-not-object-throws.js)
  - Updated according to: https://github.com/tc39/proposal-string-matchall/pull/33

peterwmwong commented on test/built-ins/RegExp/prototype/Symbol.matchAll/this-val-non-regexp.js:
```diff
@@ -1,22 +0,0 @@
-// Copyright (C) 2018 Jordan Harband. All rights reserved.
```
This test is now: [isregexp-internal-regexp-is-false.js](https://github.com/peterwmwong/test262/blob/matchall-pww/test/built-ins/RegExp/prototype/Symbol.matchAll/isregexp-internal-regexp-is-false.js)
  - Updated according to: https://github.com/tc39/proposal-string-matchall/pull/33

peterwmwong commented on test/built-ins/String/prototype/matchAll/cstm-matcher-get-err.js:
```diff
@@ -1,26 +0,0 @@
-// Copyright (C) 2018 Jordan Harband. All rights reserved.
```
This test is now:
  - [regexp-get-matchAll-throws](https://github.com/peterwmwong/test262/blob/matchall-pww/test/built-ins/String/prototype/matchAll/regexp-get-matchAll-throws.js)
  - [regexp-prototype-get-matchAll-throws](https://github.com/peterwmwong/test262/blob/matchall-pww/test/built-ins/String/prototype/matchAll/regexp-prototype-get-matchAll-throws.js)

peterwmwong commented on test/built-ins/String/prototype/matchAll/invoke-builtin-matchall.js:
```diff
@@ -1,43 +0,0 @@
-// Copyright (C) 2018 Jordan Harband. All rights reserved.
```
This test was removed as per changes from https://github.com/tc39/proposal-string-matchall/pull/33
  - Given a string, @@matchAll is no longer executed on the internally created
    RegExp.

ljharb commented on test/built-ins/RegExp/prototype/Symbol.matchAll/internal-regexp-lastindex-not-zero.js:
```diff
@@ -0,0 +1,40 @@
+    RegExp.prototype[Symbol.matchAll].call({}, '');
+  });
+} finally {
+  Object.defineProperty(RegExp.prototype, Symbol.match, {
```
Tests generally don’t need to clean up after themselves; each test is run in its own environment.

peterwmwong replied:
Great! Removed: https://github.com/ljharb/test262/pull/1/commits/7decf6aea7a44dbe4a129fa83f98d975d237e4de

ljharb commented on test/built-ins/RegExpStringIteratorPrototype/next/length.js:
```diff
@@ -0,0 +1,32 @@
+
+assert.sameValue(RegExpStringIteratorProto.next.length, 0);
+
+verifyNotEnumerable(RegExpStringIteratorProto.next, "length");
```
We should try to use consistent quotes in this pr; probably single over double?

peterwmwong replied:
Absolutely, fixed: https://github.com/ljharb/test262/pull/1/commits/0c1492d9f56d5997d06b5c7a6cb166e713d7ef04

ljharb commented on test/built-ins/RegExpStringIteratorPrototype/next/next-iteration.js:
```diff
@@ -0,0 +1,42 @@
+const str = 'a*b';
+const iter = regexp[Symbol.matchAll](str);
+
+let { value, done } = iter.next();
```
We should probably make a helper for working with iterator so; maybe giving it the iterator and an array of expected values (or an array of functions to test each value), and have it test everything? Seems like it would DRY up a lot of these tests.

peterwmwong replied:
Yeah, sounds good to me.  Is the following kinda what you had in mind?

```js
// harness/compareIterator.js
assert.compareIterator = function compareIterator(iter, validators) {
  var i = 0;
  var result = iter.next();
  while (result.done !== false) {
    validators[i](result.value);
    result = iter.next();
    i++;
  }
}

// harness/regExpUtils.js
function matchComparator(expectedEntries, expectedIndex, expectedInput) {
  return function(match) {
    assert.compareArray(match, expectedEntries);
    assert.sameValue(match.index, expectedIndex);
    assert.sameValue(match.input, expectedInput);
  }
}

// Example usage
assert.compareIterator(regexp[Symbol.matchAll](str), [
  matchComparator(['a'], 0, str),
  matchComparator(['b'], 2, str)
]);
```

ljharb replied:
Yes, that looks great! Perhaps also with something that fails if the iterator is not "done" precisely after `validators.length` iterations?

peterwmwong replied:
Ah yes! I have implemented and applied the helpers (with your suggested "done" check) in the latest commit: https://github.com/ljharb/test262/pull/1/commits/74aac5b324c37b08003943182d4be077fa7162b2

peterwmwong commented:
@ljharb I believe I've addressed your initial round of review comments:

- Remove cleanup code (undo prototype modifications)
- Consistent quoting, use single quotes
- ES5-ify: removed arrrow functions, let/const, and destructuring
- Added `assert.compareIterator` and `matchValidator`

... and am ready for another round of review!
Thank you!

peterwmwong commented:
@ljharb I believe I've addressed your initial round of review comments:

- Remove cleanup code (undo prototype modifications)
- Consistent quoting, use single quotes
- ES5-ify: removed arrrow functions, let/const, and destructuring
- Added `assert.compareIterator` and `matchValidator`

... and am ready for another round of review!
Thank you!

ljharb commented on harness/regExpUtils.js:
```diff
@@ -37,3 +37,18 @@ function testPropertyEscapes(regex, string, expression) {
+  return function(match) {
+    assert.compareArray(match, expectedEntries, 'Match entries');
+    assert.sameValue(match.index, expectedIndex, 'Match index');
+    assert.sameValue(match.input, expectedInput, 'Match input');
```
This should also take an `expectedGroups` argument, since `match` will either *lack* a `groups` property, or have one that's a null object when named capture group syntax is used.

ljharb replied:
I've rebased your branch and am going to merge it; so we can do that in a followup.
